src/group_theory/nielsen_schreier.lean

```diff
@@ -0,0 +1,334 @@
+/-
+Copyright (c) 2021 David Wärn. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: David Wärn
+-/
+import category_theory.action
+import category_theory.is_connected
+import combinatorics.quiver
+import group_theory.free_group
+import group_theory.semidirect_product
+/-!
+# The Nielsen-Schreier theorem
+
+This file proves that a subgroup of a free group is itself free.
+
+## Main definitions
+
+- `is_free_group G`: a class expressing that `G` has the universal property of a free group.
+- `subgroup_is_free_of_free H`: an instance saying that a subgroup of a free group is free.
+
+## Proof overview
+
+The proof is analogous to the proof using covering spaces and fundamental groups of graphs,
+but we work directly with groupoids instead of topological spaces. Under this analogy,
+
+- `is_free_groupoid G` corresponds to saying that a space is a graph.
+- `End_mul_equiv_subgroup H` plays the role of replacing 'subgroup of fundamental group' with
+  'fundamental group of covering space'.
+- `action_category_is_free G A` corresponds to the fact that a covering of a (single-vertex)
+  graph is a graph.
+- `End_is_free_group_of_arborescence` corresponds to the fact that, given a spanning tree of a
+  graph, its fundamental group is free (generated by loops from the complement of the tree).
+
+## Implementation notes
+
+This proof works with the universal property `is_free_group` instead of the concrete description
+`free_group A` of free groups. The two are related by `free_group_is_free_group` and
+`iso_free_group_of_is_free_group`.
+
+Our definition of `is_free_groupoid` is nonstandard. Normally one would require that functors
+`G ⥤ X` to any _groupoid_ `X` are given by graph homomorphisms from the generators, but we only
+consider _groups_ `X`. This simplifies the argument since functor equality is complicated in
+general, but simple for functors to single object categories.
+
+## References
+
+https://ncatlab.org/nlab/show/Nielsen-Schreier+theorem
+
+## Tags
+
+free group, Nielsen-Schreier
+
+-/
+
+noncomputable theory
+open_locale classical
+universes v u
+
+open category_theory category_theory.action_category semidirect_product
+
+/-- `is_free_group G` means that `G` has the universal property of a free group.
+    That is, it has a family `generators G` of elements, such that a group homomorphism
+    `G →* X` is uniquely determined by a function `generators G → X`. -/
+class is_free_group (G) [group.{u} G] :=
+(generators : Type u)
+(of : generators → G)
+(unique_lift : ∀ {X} [group.{u} X] (f : generators → X),
+                ∃! F : G →* X, ∀ a, F (of a) = f a)
```

Comment:
What would you think of moving this, and material about it before you really embark on Nielsen-Schreier, to its own file? Similarly for `is_free_groupoid`?

Reply:
We could do this, but I'm not sure `is_free_group(oid)` would be useful outside this context. There is already `free_group` of course. I think of `is_free_group` as mostly just guiding this proof -- it could be written entirely in terms of `free_group`, but that would only make the proof more opaque.

Reply:
I suspect `is_free_group` might actually be quite useful. There are many appearances of free groups which are not "explicit". For example to prove the braid group acts on the free group by automorphisms it's helpful to think of "the free group" instead as a fundamental group. Or often we want to prove that some group contains a subgroup which is free on some number of generators. For either of these applications `is_free_group` seems useful.

Reply:
I'll second what Scott said in the above comment. Since this definition of `is_free_group` restricts the universe of `X` and `generators` (which is good!), I think it would be useful to add a lemma analogous to `unique_lift` which is universe polymorphic. This should be straightforward since you already have the isomorphism with "the" free group in this file.